File: src/types/domain.ts
// ── Legacy types (kept for migration compatibility) ──────────────────────────
export type RoomStatus =
  | "CHECKOUT"
  | "PENDING"
  | "CLEANING"
  | "READY"
  | "AVAILABLE"
  | "MAINTENANCE";

export type Property = {
  id: string;
  code: string;
  name: string;
  totalRooms: number;
};

// Legacy — still used by roomsService / mockRooms for seeding
export type RoomCard = {
  id: string;
  propertyId: string;
  roomCode: string;
  title: string;
  status: RoomStatus;
  assignedTo?: string;
  cleanedBy?: string;
  cleanedAt?: string;
  notes?: string;
  priority?: "LOW" | "NORMAL" | "HIGH";
  hasNextDayCheckin?: boolean;
  checkinDate?: string;
};

export type MockUser = {
  id: string;
  name: string;
  initials: string;
};

// Legacy — kept for ArchivePage
export type ArchivedRoom = RoomCard & {
  archivedAt: string;
  archivedBy: string;
};

export type PropertyTaskCategory = "AVISO" | "COMPRAS";
export type PropertyTask = {
  id: string;
  propertyId: string;
  category: PropertyTaskCategory;
  text: string;
  done: boolean;
  createdAt: string;
};

// ── New dynamic board types ───────────────────────────────────────────────────

export type ColumnColor =
  | "green" | "yellow" | "orange" | "red"
  | "purple" | "blue" | "cyan" | "lime" | "pink" | "gray";

export const COLUMN_COLOR_CLASSES: Record<
  ColumnColor,
  { border: string; headerBg: string; iconColor: string; swatch: string; dot: string }
> = {
  green:  { border: "border-t-green-400",   headerBg: "bg-green-50",   iconColor: "text-green-600",   swatch: "bg-green-400",   dot: "bg-green-500"   },
  yellow: { border: "border-t-amber-400",   headerBg: "bg-amber-50",   iconColor: "text-amber-600",   swatch: "bg-amber-400",   dot: "bg-amber-500"   },
  orange: { border: "border-t-orange-400",  headerBg: "bg-orange-50",  iconColor: "text-orange-600",  swatch: "bg-orange-400",  dot: "bg-orange-500"  },
  red:    { border: "border-t-red-400",     headerBg: "bg-red-50",     iconColor: "text-red-600",     swatch: "bg-red-400",     dot: "bg-red-500"     },
  purple: { border: "border-t-purple-400",  headerBg: "bg-purple-50",  iconColor: "text-purple-600",  swatch: "bg-purple-400",  dot: "bg-purple-500"  },
  blue:   { border: "border-t-blue-400",    headerBg: "bg-blue-50",    iconColor: "text-blue-600",    swatch: "bg-blue-400",    dot: "bg-blue-500"    },
  cyan:   { border: "border-t-cyan-400",    headerBg: "bg-cyan-50",    iconColor: "text-cyan-600",    swatch: "bg-cyan-400",    dot: "bg-cyan-500"    },
  lime:   { border: "border-t-lime-400",    headerBg: "bg-lime-50",    iconColor: "text-lime-600",    swatch: "bg-lime-400",    dot: "bg-lime-500"    },
  pink:   { border: "border-t-pink-400",    headerBg: "bg-pink-50",    iconColor: "text-pink-600",    swatch: "bg-pink-400",    dot: "bg-pink-500"    },
  gray:   { border: "border-t-slate-400",   headerBg: "bg-slate-50",   iconColor: "text-slate-600",   swatch: "bg-slate-400",   dot: "bg-slate-500"   },
};

export const COLUMN_COLORS_ORDER: ColumnColor[] = [
  "green", "yellow", "orange", "red", "purple",
  "blue",  "cyan",   "lime",   "pink", "gray",
];

/** Per-column visual configuration — all booleans default to true when undefined */
export type ColumnConfig = {
  // ── Card modal sections ──────────────────────────────────────────────────
  showDescription?: boolean;    // show Description section in card modal
  showCustomFields?: boolean;   // show Custom Fields section in card modal
  showComments?: boolean;       // show Comments/Activity section in card modal
  // ── Card modal sidebar ───────────────────────────────────────────────────
  showAssign?: boolean;         // show Asignar sidebar action
  showPriority?: boolean;       // show Prioridad sidebar action
  showCheckin?: boolean;        // show Fecha checkin sidebar action
  // ── Card item display ────────────────────────────────────────────────────
  cardChipFieldIds?: string[];  // which field chips show on cards (undefined = use global showOnCard)
  showCardDescription?: boolean; // description preview line on card
  showCardRoomCode?: boolean;    // room code badge on card
  showCardFooter?: boolean;      // comments / date / assignee footer icons
  showCardDoneStamp?: boolean;   // green "cleaned by" done stamp
};

export type Column = {
  id: string;
  propertyId: string;
  title: string;
  color: ColumnColor;
  position: number;
  archived?: boolean;
  description?: string;
  /** If set, only these field IDs are shown in card modals for this column.
   *  If undefined, all globally enabled fields are shown. */
  fieldIds?: string[];
  /** Per-column visual configuration for cards and card modal */
  columnConfig?: ColumnConfig;
};

export type CardKind = "ROOM" | "TASK" | "FREE";

// Flexible custom fields — keys map to FieldDef ids
export type CardCustomFields = Record<string, string | boolean | undefined>;

// Field definition for settings
export type FieldType = "text" | "select" | "checkbox" | "image";

export type FieldDef = {
  id: string;
  label: string;
  type: FieldType;
  options?: string[];     // for select
  enabled: boolean;       // show in card modal
  showOnCard?: boolean;   // show as chip on CardItem
  icon?: string;          // mdi icon name
  offLabel?: string;      // for checkbox: label shown when unchecked (always visible if set)
};

export type Card = {
  id: string;
  propertyId: string;
  columnId: string;
  title: string;
  description?: string;
  isDone: boolean;
  position: number;
  createdAt: string;
  kind: CardKind;
  customFields?: CardCustomFields;
  // ROOM-specific
  roomCode?: string;
  cleanedBy?: string;
  cleanedAt?: string;
  priority?: "LOW" | "NORMAL" | "HIGH";
  hasNextDayCheckin?: boolean;
  checkinDate?: string;
  assignedTo?: string;
  notes?: string;
};

export type ArchivedCard = Card & {
  archivedAt: string;
  archivedBy: string;
};

export type Comment = {
  id: string;
  cardId: string;
  author: string;
  authorInitials: string;
  text: string;
  createdAt: string;
};

export type ActivityEvent = {
  id: string;
  cardId: string;
  type: "created" | "moved" | "archived" | "done" | "undone" | "commented";
  message: string;
  createdAt: string;
};

File: manage_property_room_api/internal/domain/domain.go
// Package domain holds pure entities. Mirrors the Flutter `lib/domain` layer.
package domain

import "time"

// ── Enums (string-typed for JSON parity with Flutter `.name`) ──

type UserRole string

const (
	RoleAdmin       UserRole = "admin"
	RoleOperator    UserRole = "operator"
	RoleCleaning    UserRole = "cleaning"
	RoleMaintenance UserRole = "maintenance"
)

type CardKind string

const (
	CardKindRoom CardKind = "room"
	CardKindTask CardKind = "task"
	CardKindFree CardKind = "free"
)

type FieldType string

const (
	FieldText     FieldType = "text"
	FieldSelect   FieldType = "select"
	FieldCheckbox FieldType = "checkbox"
	FieldImage    FieldType = "image"
)

type ColumnColor string

const (
	ColorGreen  ColumnColor = "green"
	ColorYellow ColumnColor = "yellow"
	ColorOrange ColumnColor = "orange"
	ColorRed    ColumnColor = "red"
	ColorPurple ColumnColor = "purple"
	ColorBlue   ColumnColor = "blue"
	ColorCyan   ColumnColor = "cyan"
	ColorLime   ColumnColor = "lime"
	ColorPink   ColumnColor = "pink"
	ColorGray   ColumnColor = "gray"
)

type CardPriority string

const (
	PriorityLow    CardPriority = "low"
	PriorityNormal CardPriority = "normal"
	PriorityHigh   CardPriority = "high"
)

type ActivityType string

const (
	ActivityCreated   ActivityType = "created"
	ActivityMoved     ActivityType = "moved"
	ActivityArchived  ActivityType = "archived"
	ActivityDone      ActivityType = "done"
	ActivityUndone    ActivityType = "undone"
	ActivityCommented ActivityType = "commented"
	ActivityAssigned  ActivityType = "assigned"
	ActivityEdited    ActivityType = "edited"
)

type ArchiveKind string

const (
	ArchiveKindCard   ArchiveKind = "card"
	ArchiveKindColumn ArchiveKind = "column"
)

// ── Entities ──

type User struct {
	ID                  string    `json:"id" db:"id"`
	Email               string    `json:"email" db:"email"`
	PasswordHash        string    `json:"-" db:"password_hash"`
	Name                string    `json:"name" db:"name"`
	Initials            string    `json:"initials" db:"initials"`
	Role                UserRole  `json:"role" db:"role"`
	MustChangePassword  bool      `json:"mustChangePassword" db:"must_change_password"`
	AssignedPropertyIDs []string  `json:"assignedPropertyIds" db:"-"`
	CreatedAt           time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt           time.Time `json:"updatedAt" db:"updated_at"`
}

type Property struct {
	ID         string    `json:"id" db:"id"`
	Code       string    `json:"code" db:"code"`
	Name       string    `json:"name" db:"name"`
	TotalRooms int       `json:"totalRooms" db:"total_rooms"`
	ColorSeed  int       `json:"colorSeed" db:"color_seed"`
	Position   int       `json:"position" db:"position"`
	Archived   bool      `json:"archived" db:"archived"`
	CreatedAt  time.Time `json:"createdAt" db:"created_at"`
	UpdatedAt  time.Time `json:"updatedAt" db:"updated_at"`
}

type ColumnConfig struct {
	// Form-level toggles (shown in card detail sheet)
	ShowDescription  *bool `json:"showDescription,omitempty"`
	ShowCustomFields *bool `json:"showCustomFields,omitempty"`
	ShowComments     *bool `json:"showComments,omitempty"`
	ShowAssign       *bool `json:"showAssign,omitempty"`
	ShowPriority     *bool `json:"showPriority,omitempty"`
	ShowCheckin      *bool `json:"showCheckin,omitempty"`
	// Card chip field IDs
	CardChipFieldIDs []string `json:"cardChipFieldIds,omitempty"`
	// Card compact-view toggles (persistent, formerly CardDisplayPrefs)
	CardShowDone           *bool `json:"cardShowDone,omitempty"`
	CardShowDescription    *bool `json:"cardShowDescription,omitempty"`
	CardShowCleanedBy      *bool `json:"cardShowCleanedBy,omitempty"`
	CardShowPriority       *bool `json:"cardShowPriority,omitempty"`
	CardShowPriorityBorder *bool `json:"cardShowPriorityBorder,omitempty"`
	CardShowCheckin        *bool `json:"cardShowCheckin,omitempty"`
	CardShowRoomCode       *bool `json:"cardShowRoomCode,omitempty"`
	CardShowImage          *bool `json:"cardShowImage,omitempty"`
}

type BoardColumn struct {
	ID          string        `json:"id" db:"id"`
	PropertyID  string        `json:"propertyId" db:"property_id"`
	Title       string        `json:"title" db:"title"`
	Color       ColumnColor   `json:"color" db:"color"`
	Position    int           `json:"position" db:"position"`
	Description string        `json:"description" db:"description"`
	Archived    bool          `json:"archived" db:"archived"`
	FieldIDs    []string      `json:"fieldIds,omitempty" db:"-"`
	Config      *ColumnConfig `json:"columnConfig,omitempty" db:"-"`
	CreatedAt   time.Time     `json:"createdAt" db:"created_at"`
	UpdatedAt   time.Time     `json:"updatedAt" db:"updated_at"`
}

type Card struct {
	ID           string                 `json:"id" db:"id"`
	PropertyID   string                 `json:"propertyId" db:"property_id"`
	ColumnID     string                 `json:"columnId" db:"column_id"`
	Title        string                 `json:"title" db:"title"`
	Description  string                 `json:"description" db:"description"`
	Position     int                    `json:"position" db:"position"`
	IsDone       bool                   `json:"isDone" db:"is_done"`
	Archived     bool                   `json:"archived" db:"archived"`
	RoomCode     string                 `json:"roomCode" db:"room_code"`
	Priority     CardPriority           `json:"priority" db:"priority"`
	CheckinDate  *time.Time             `json:"checkinDate,omitempty" db:"checkin_date"`
	AssignedToID *string                `json:"assignedToId,omitempty" db:"assigned_to_id"`
	Kind         CardKind               `json:"kind" db:"kind"`
	CustomFields map[string]any         `json:"customFields" db:"-"`
	CreatedAt    time.Time              `json:"createdAt" db:"created_at"`
	UpdatedAt    time.Time              `json:"updatedAt" db:"updated_at"`
	CleanedBy    string                 `json:"cleanedBy" db:"cleaned_by"`
	DoneAt       *time.Time             `json:"doneAt,omitempty" db:"done_at"`
}

type FieldDef struct {
	ID         string    `json:"id" db:"id"`
	Label      string    `json:"label" db:"label"`
	Type       FieldType `json:"type" db:"type"`
	Options    []string  `json:"options,omitempty" db:"-"`
	Enabled    bool      `json:"enabled" db:"enabled"`
	ShowOnCard bool      `json:"showOnCard" db:"show_on_card"`
	Icon       string    `json:"icon,omitempty" db:"icon"`
	OffLabel   string    `json:"offLabel,omitempty" db:"off_label"`
	Position   int       `json:"position" db:"position"`
}

type Comment struct {
	ID        string    `json:"id" db:"id"`
	CardID    string    `json:"cardId" db:"card_id"`
	AuthorID  string    `json:"authorId" db:"author_id"`
	Text      string    `json:"text" db:"text"`
	CreatedAt time.Time `json:"createdAt" db:"created_at"`
}

type ActivityEvent struct {
	ID        string         `json:"id" db:"id"`
	CardID    string         `json:"cardId" db:"card_id"`
	UserID    string         `json:"userId" db:"user_id"`
	Type      ActivityType   `json:"type" db:"type"`
	Payload   map[string]any `json:"payload,omitempty" db:"-"`
	CreatedAt time.Time      `json:"createdAt" db:"created_at"`
}

type ArchiveItem struct {
	ID         string         `json:"id" db:"id"`
	Kind       ArchiveKind    `json:"kind" db:"kind"`
	Payload    map[string]any `json:"payload" db:"-"`
	ArchivedAt time.Time      `json:"archivedAt" db:"archived_at"`
}

// AuditEvent records a mutation performed by a user.
type AuditEvent struct {
	ID        string    `json:"id" db:"id"`
	ActorID   string    `json:"actorId" db:"actor_id"`
	ActorName string    `json:"actorName" db:"actor_name"`
	Action    string    `json:"action" db:"action"`
	Entity    string    `json:"entity" db:"entity"`
	EntityID  string    `json:"entityId" db:"entity_id"`
	Detail    string    `json:"detail,omitempty" db:"detail"`
	CreatedAt time.Time `json:"createdAt" db:"created_at"`
}
